### PR TITLE
Scalar compile func

### PR DESCRIPTION
**Pull request by andyfengHKU**
Add scalar_compile_func to remove special function evaluation that only happens at compile time.

Add `nodes(RECURSIVE_REL), reps(RECURSIVE_REL)`

### CODE REVIEW

File: src/binder/bind_expression/bind_function_expression.cpp [hunk 1]
```diff
@@ -67,9 +67,10 @@ std::shared_ptr<Expression> ExpressionBinder::bindScalarFunctionExpression(
     }
     auto uniqueExpressionName =
         ScalarFunctionExpression::getUniqueName(function->name, childrenAfterCast);
-    return make_shared<ScalarFunctionExpression>(functionName, FUNCTION, std::move(bindData),
+    auto result = make_shared<ScalarFunctionExpression>(functionName, FUNCTION, std::move(bindData),
         std::move(childrenAfterCast), function->execFunc, function->selectFunc,
-        uniqueExpressionName);
+        function->compileFunc, uniqueExpressionName);
+    return result;
```
Comment (andyfengHKU) [inline src/binder/bind_expression/bind_function_expression.cpp:73]: revert

File: src/include/function/vector_operations.h [hunk 2]
```diff
@@ -12,6 +12,9 @@ namespace function {
 
 struct VectorOperationDefinition;
 
+using scalar_compile_func =
+    std::function<void(FunctionBindData*, const std::vector<std::shared_ptr<common::ValueVector>>&,
+        std::shared_ptr<common::ValueVector>&)>;
```
Comment (acquamarin) [inline src/include/function/vector_operations.h:17]: i think we never use a reference to a shared_ptr, either change it to a raw ptr or remove reference

File: src/function/vector_union_operations.cpp [hunk 3]
```diff
@@ -34,12 +35,22 @@ void UnionValueVectorOperations::execFunc(
     common::UnionVector::setTagField(&result, 0 /* tagFieldIdx */);
 }
 
+void UnionValueVectorOperations::compileFunc(FunctionBindData* bindData,
+    const std::vector<std::shared_ptr<common::ValueVector>>& parameters,
+    std::shared_ptr<common::ValueVector>& result) {
+    assert(parameters.size() == 1);
+    // TODO(Ziyi): this may change as we try to create union value given different children states.
```
Comment (acquamarin) [inline src/function/vector_union_operations.cpp:42]: As we discussed, let's remove this todo.

File: src/expression_evaluator/function_evaluator.cpp [hunk 4]
```diff
@@ -16,19 +15,13 @@ void FunctionExpressionEvaluator::init(const ResultSet& resultSet, MemoryManager
...
-    for (auto& child : children) {
-        parameters.push_back(child->resultVector);
-    }
 }
 
 void FunctionExpressionEvaluator::evaluate() {
     for (auto& child : children) {
         child->evaluate();
     }
-    if (execFunc != nullptr) {
```
Comment (acquamarin) [inline src/expression_evaluator/function_evaluator.cpp:28]: up to you, i think our compile/bind function can be a nullptr, better to make is consistent with these?

**Review by acquamarin — APPROVED** [review]

File: src/include/function/vector_operations.h (hunk 2 above)
Comment (andyfengHKU) [reply src/include/function/vector_operations.h:17]: This has to be a reference, because we are changing the shared_ptr inside compileFunc

**Pull request merged**